### PR TITLE
Max bipartite matching

### PR DESCRIPTION
**Pull request by eric7237cire**
Another PR adding a test that shows how to use the max flow to find a maximum matching.

I had to refactor the dinic method a bit to return the flow.

### CODE REVIEW

File: src/graph/flow.rs [hunk 1]
```diff
@@ -43,6 +43,10 @@ impl FlowGraph {
     ///
     /// Panics if the maximum flow is 2^63 or larger.
     pub fn dinic(&self, s: usize, t: usize) -> i64 {
```
Comment (EbTech) [inline src/graph/flow.rs]: I'd like to avoid the extra function.
Feel free to modify both dinic() and mcf() to return tuples that include the flow.

File: src/graph/flow.rs [hunk 2]
```diff
@@ -221,4 +225,53 @@ mod test {
...
+        graph.add_edge(a_start + 4, b_start + 2, 1, 1);
+        graph.add_edge(a_start + 4, b_start + 3, 1, 1);
+        graph.add_edge(a_start + 5, b_start + 5, 1, 1);
+
+        let (flow_amt, flow) = graph.dinic_flow(source, sink);
+        assert_eq!(flow_amt, 5);
+
+        //U->V edges in maximum matching
+        assert_eq!(
+            flow
```
Comment (EbTech) [inline src/graph/flow.rs]: Nice! Consider giving this expression a variable name :P

Comment (EbTech) [reply src/graph/flow.rs]: Also consider using into_iter() which consumes flow and lets you use f instead of *f

File: src/graph/flow.rs (hunk 1 above)
Comment (eric7237cire) [reply src/graph/flow.rs]: ok, changed to use 1 signature for both mcf & dinic

File: src/graph/flow.rs (hunk 2 above)
Comment (eric7237cire) [reply src/graph/flow.rs]: A nice tip, changed.

**Comment by eric7237cire** [comment]
This PR actually has the requested changes for PR #5 so merging this one will include both.  I hadn't branched the other one.  Sorry about that.

File: src/graph/dfs.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            // if it is not visited.
+            if !self.visited[s] {
+                self.visited.set(s, true);
+                r = Some(s);
+
+                // Get all adjacent vertices of the popped vertex s
+                // If a adjacent has not been visited, then puah it
+                // to the stack.
+                for (_e, u) in self.graph.adj_list(s) {
+                    self.stack.push(u);
```
[the remark this answers is not in the document]
Comment (eg-novelt) [reply src/graph/dfs.rs]: Ah ooops.  The wikipedia page says it has space O(|e|), which in a complete graph is approaching V^2.  As an aside, its interesting how recursive functions can be so tricky to make iterative.

I took a crack at it and added a test as well to make sure it handles the complete graph case without blowing up the stack.

File: src/graph/dfs.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    //stack of vertices
+    stack: Vec<usize>,
+}
+
+impl<'a> Iterator for DfsIterator<'a> {
+    type Item = usize;
+
+    /// Returns next vertex in the DFS
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut r = None;
```
[the remark this answers is not in the document]
Comment (eg-novelt) [reply src/graph/dfs.rs]: After I put in the adj iterators, I still needed it.  If you can find a way to get it out of there, please let me know.

File: src/graph/dfs.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        while let Some(s) = self.stack.pop() {
+            // Stack may contain same vertex twice. So
+            // we need to print the popped item only
+            // if it is not visited.
+            if !self.visited[s] {
+                self.visited.set(s, true);
+                r = Some(s);
+
+                // Get all adjacent vertices of the popped vertex s
+                // If a adjacent has not been visited, then puah it
```
[the remark this answers is not in the document]
Comment (eg-novelt) [reply src/graph/dfs.rs]: That comment is gone/changed now

File: src/graph/dfs.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (eg-novelt) [reply src/graph/dfs.rs]: r is gone

File: src/graph/dfs.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        graph.add_edge(4, 0);
+
+        let dfs_search = graph.dfs(0).collect::<Vec<_>>();
+        //Note this is not the only valid DFS
+        assert_eq!(dfs_search, vec![0, 3, 4, 2, 1]);
+    }
+
+    #[test]
+    fn test_dfs_space_complexity() {
+        let mut graph = Graph::new(20, 0);
```
Comment (EbTech) [inline src/graph/dfs.rs]: Since the number 20 appears many times, consider declaring it once:
let num_v = 20

File: src/graph/dfs.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    /// Returns next vertex in the DFS
+    fn next(&mut self) -> Option<Self::Item> {
+        //Sources:
+        // https://www.geeksforgeeks.org/iterative-depth-first-traversal/
+        // https://en.wikipedia.org/wiki/Depth-first_search
+        while let Some(s) = self.stack.last() {
+            let s = *s;
+
+            //Does s still have neighbors we need to process?
+            if let Some(s_nbr) = self.adj_iters[s].next() {
```
Comment (EbTech) [inline src/graph/dfs.rs]: Some((_, v)) can be used to extract the vertex directly

File: src/graph/dfs.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+impl<'a> Iterator for DfsIterator<'a> {
+    type Item = usize;
+
+    /// Returns next vertex in the DFS
+    fn next(&mut self) -> Option<Self::Item> {
+        //Sources:
+        // https://www.geeksforgeeks.org/iterative-depth-first-traversal/
+        // https://en.wikipedia.org/wiki/Depth-first_search
+        while let Some(s) = self.stack.last() {
```
Comment (EbTech) [inline src/graph/dfs.rs]: Some(&s) can be used to dereference directly

Comment (eric7237cire) [reply src/graph/dfs.rs]: Ah yes, that's cleaner.  Done

File: src/graph/dfs.rs (hunk 6 above)
Comment (eric7237cire) [reply src/graph/dfs.rs]: Agreed, done

File: src/graph/dfs.rs (hunk 7 above)
Comment (eric7237cire) [reply src/graph/dfs.rs]: Yes, it is cleaner that way, done

**Comment by eric7237cire** [comment]
> Cool, that should work!

Yeah I was pretty surprised the iterative DFS on the wikipedia page wasn't the most space efficient.  Maybe because its a touch trickier to implement.  Who knows.

**Comment by EbTech** [comment]
Awesome, I merged it in. Thanks for your contribution!
